contracts/pcv/PCVGuardian.sol

```diff
@@ -0,0 +1,185 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+pragma solidity ^0.8.4;
+
+import {EnumerableSet} from "@openzeppelin/contracts/utils/structs/EnumerableSet.sol";
+import {CoreRef} from "../refs/CoreRef.sol";
+import {IPCVGuardian} from "./IPCVGuardian.sol";
+import {IPCVDeposit} from "./IPCVDeposit.sol";
+import {CoreRefPauseableLib} from "../libs/CoreRefPauseableLib.sol";
+import {TribeRoles} from "../core/TribeRoles.sol";
+
+/// @notice PCV Guardian is a contract to safeguard protocol funds
+/// by being able to withdraw whitelisted PCV deposits to an immutable safe address
+contract PCVGuardian is IPCVGuardian, CoreRef {
+    using CoreRefPauseableLib for address;
+    using EnumerableSet for EnumerableSet.AddressSet;
+
+    ///@notice set of whitelisted pcvDeposit addresses for withdrawal
+    EnumerableSet.AddressSet private whitelistAddresses;
+
+    ///@notice safe address where PCV funds can be withdrawn to
+    address public immutable safeAddress;
+
+    constructor(
+        address _core,
+        address _safeAddress,
+        address[] memory _whitelistAddresses
+    ) CoreRef(_core) {
+        safeAddress = _safeAddress;
+
+        // improbable to ever overflow
+        unchecked {
+            for (uint256 i = 0; i < _whitelistAddresses.length; i++) {
+                _setWhitelistAddress(_whitelistAddresses[i]);
+            }
+        }
+    }
+
+    // ---------- Read-Only API ----------
+
+    /// @notice returns true if the the provided address is a valid destination to withdraw funds from
+    /// @param pcvDeposit the pcvDeposit address to check if whitelisted
+    function isWhitelistAddress(address pcvDeposit)
+        public
+        view
+        override
+        returns (bool)
+    {
+        return whitelistAddresses.contains(pcvDeposit);
+    }
+
+    /// @notice returns all whitelisted pcvDeposit addresses
+    function getWhitelistAddresses()
+        public
+        view
+        override
+        returns (address[] memory)
+    {
+        return whitelistAddresses.values();
+    }
+
+    // ---------- Governor-Only State-Changing API ----------
+
+    /// @notice governor-only method to whitelist a pcvDesposit address to withdraw funds from
+    /// @param pcvDeposit the address to set as safe
+    function setWhitelistAddress(address pcvDeposit)
+        external
+        override
+        onlyGovernor
+    {
+        _setWhitelistAddress(pcvDeposit);
+    }
+
+    /// @notice batch version of setWhiteListaddress
+    /// @param _whitelistAddresses the addresses to whitelist, as calldata
+    function setWhitelistAddresses(address[] calldata _whitelistAddresses)
```

Comment:
what's the difference is gas calling this with a single address vs calling the `setWhitelistAddress(address)` method?

Reply:
if you have to do 5 tx's to set the whitelist address without this bulk whitelisting method, then it would cost you at least 21k gas * 4 tx's as the base tx validation cost is 21k gas, so its at least 84k gas more expensive not factoring in the cost increase of using the stack/memory instead of calldata.

Reply:
Oops, I meant the other way (for example if `setWhitelistAddress` was removed and only `setWhitelistAddresses` remained)

Reply:
Was curious so tested in remix 
<img width="892" alt="image" src="https://user-images.githubusercontent.com/15036618/167674703-c566a992-46ca-4d48-a371-a2f1d80b7e3e.png">

You could maybe remove the `setWhitelistAddress` method depending on how you plan on adding pcvDeposits. Doesn't really matter tho